%% LyX 1.3 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[english, 12pt]{article}
\usepackage{times}
%\usepackage{algorithm2e}
\usepackage{url}
\usepackage{bbm}
\usepackage[T1]{fontenc}
\usepackage[latin1]{inputenc}
\usepackage{geometry}
\geometry{verbose,letterpaper,tmargin=2cm,bmargin=2cm,lmargin=1.5cm,rmargin=1.5cm}
\usepackage{rotating}
\usepackage{color}
\usepackage{graphicx}
\usepackage{amsmath, amsthm, amssymb}
\usepackage{setspace}
\usepackage{lineno}
\usepackage{hyperref}
\usepackage{bbm}
\usepackage{makecell}

\renewcommand{\arraystretch}{1.3}

%\usepackage{xr}
%\externaldocument{240pgs-supp}

%\linenumbers
%\doublespacing
\onehalfspacing
%\usepackage[authoryear]{natbib}
\usepackage{natbib} \bibpunct{(}{)}{;}{author-year}{}{,}

%Pour les rajouts
\usepackage{color}
\definecolor{trustcolor}{rgb}{0,0,1}

\usepackage{dsfont}
\usepackage[warn]{textcomp}
\usepackage{adjustbox}
\usepackage{multirow}
\usepackage{graphicx}
\graphicspath{{../figures/}}
\DeclareMathOperator*{\argmin}{\arg\!\min}
\usepackage{algorithm}
\usepackage{algpseudocode}

\let\tabbeg\tabular
\let\tabend\endtabular
\renewenvironment{tabular}{\begin{adjustbox}{max width=0.9\textwidth}\tabbeg}{\tabend\end{adjustbox}}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Bold symbol macro for standard LaTeX users
%\newcommand{\boldsymbol}[1]{\mbox{\boldmath $#1$}}

%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}

\usepackage{babel}
\makeatother


\begin{document}


\title{Application Note\\lassosum2: an updated version complementing LDpred2}
\author{Florian Priv\'e,$^{\text{1,}*}$ Bjarni J. Vilhj\'almsson,$^{\text{1,2}}$ Shing Wan Choi,$^{\text{3}}$ and Timothy Shin Heng Mak$^{\text{4}}$}

\date{~ }
\maketitle

\noindent$^{\text{\sf 1}}$National Centre for Register-Based Research, Aarhus University, Aarhus, 8210, Denmark. \\
\noindent$^{\text{\sf 2}}$Bioinformatics Research Centre, Aarhus University, Aarhus, 8000, Denmark. \\
\noindent$^{\text{\sf 3}}$Department of Genetics and Genomic Sciences, Icahn School of Medicine at Mount Sinai, New York, New York, 10029, USA \\
\noindent$^{\text{\sf 4}}$[TO FILL] \\
\noindent$^\ast$To whom correspondence should be addressed.\\

\noindent Contact: \url{florian.prive.21@gmail.com}

\vspace*{8em}

\abstract{
  We present lassosum2, a new version of the polygenic score method lassosum. This is now implemented in package bigsnpr and uses the exact same input data as LDpred2. This new version is also very fast, which means that it can be run with almost no extra coding nor computational time when already running LDpred2. It can also be more robust than LDpred2, e.g.\ when there is some large GWAS sample size misspecification, making lassosum2 complementary to LDpred2.
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\clearpage

\section*{Introduction}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{New implementation}

From \cite{mak2017polygenic}, the solution from lassosum can be obtained by iteratively updating 
\[
\beta_j^{(t)} =
\begin{cases}
\text{sign}\left(u_j^{(t)}\right) \left|u_j^{(t)} - \lambda\right| / \left(\widetilde{X}_j^T \widetilde{X}_j + s\right) & \text{if } \left|u_j^{(t)}\right| > \lambda ~, \\
0 & \text{otherwise.}
\end{cases}
\]
where 
\[
u_j^{(t)} = r_j - \widetilde{X}_j^T \left( \widetilde{X} \beta^{(t-1)} - \widetilde{X}_j \beta_j^{(t-1)} \right) ~.
\]
Following notations in \cite{prive2020ldpred2} and denoting $\widetilde{X} = \sqrt{\frac{1-s}{n-1}} C_n G S^{-1}$, then $\widetilde{X}_j^T \widetilde{X} = (1-s) R_{j,.} = (1-s) R_{.,j}^T$ where $R$ is the correlation matrix between variants.
Then $\widetilde{X}_j^T \widetilde{X}_j + s = 1$ and
\[
u_j^{(t)} = \widehat{\beta}_j + (1-s) \left( \beta_j^{(t-1)} - R_{.,j}^T \beta^{(t-1)} \right) ~,
\]
where $r_j = \widehat{\beta}_j =  \dfrac{\widehat{\gamma}_j}{\sqrt{n_j ~ \text{se}(\widehat{\gamma}_j)^2 + \widehat{\gamma}_j^2}}$ and $\widehat{\gamma}_j$ is the effect of variant $j$ from the GWAS \cite[]{mak2017polygenic,prive2021high}.
Then the most time-consuming part is computing $R_{.,j}^T \beta^{(t-1)}$.
To make this faster, instead of computing $R_{.,j}^T \beta^{(t-1)}$ at each iteration, we can start with a vector with only 0s initially (for all $j$) since $\beta^{(0)} \equiv 0$, and then updating this vector when $\beta_j^{(t)} \neq \beta_j^{(t-1)}$ only. Note that only positions $k$ for which $R_{k,j} \neq 0$ must be updated in this vector. 

In this new implementation of the lassosum model, the input parameters are the correlation matrix $R$, the GWAS summary statistics ($\widehat{\gamma}_j$, $\text{se}(\widehat{\gamma}_j)$ and $n_j$), and the two hyper-parameters $\lambda$ and $s$. 
Therefore, except from the hyper-parameters, this is the exact same input as for LDpred2 \cite[]{prive2020ldpred2}.
For $s$, we now try $\{0.1, 0.2, \dots, 1.0\}$ by default instead of only $\{0.2, 0.5, 0.8, 1.0\}$.
For $\lambda$, the default in lassosum uses a sequence of 20 values equally spaced on a log scale between 0.1 and 0.001; we now use a sequence between $\lambda_0$ and $\lambda_0 / 100$ by default in lassosum2, where $\lambda_0 = \max_j \left|\widehat{\beta}_j\right|$ is the minimum $\lambda$ for which no variable enter the model because the L1-regularization is too strong.
Note that we do not provide an ``auto'' version using pseudo-validation as in \cite{mak2017polygenic} as we have not found it very robust [TODO: ADD FIGURE SHOWING THIS + FIG \ref{fig:auc}].
Also note that, as in LDpred2, we run lassosum2 genome-wide and do not require splitting the genome into independent LD blocks.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Results}

We use the exact same data as in \cite{prive2020ldpred2}; we use external summary statistics for eight case-control phenotypes and predict in the UK Biobank (UKBB) European data by keeping 10,000 individuals as validation set for tuning hyper-parameters and using the remaining 352,320 individuals as test set.
As we use the exact same split as before, results can be directly compared to results in \cite{prive2020ldpred2} and we run only lassosum2 as the new method.
Figure \ref{fig:auc} compares lassosum2 with lassosum and LDpred2. AUC values for lassosum2 are similar to lassosum for breast cancer (BRCA), coronary artery disease (CAD), depression (MDD), prostate cancer (PRCA), rheumatoid arthritis (RA) and type 2 diabetes (T2D), and slighly better for asthma and type 1 diabetes (T1D).
Prediction with LDpred2 is still slightly higher than with lassosum2, except for MDD and TAD for which they are similar (Figure \ref{fig:auc}).

\begin{figure}[htb]
	\centerline{\includegraphics[width=0.85\textwidth]{AUC-lassosum}}
	\caption{lassosum2 is compared with LDpred2, LDpred2-auto, lassosum and lassosum-auto using  published  external  summary  statistics and the UKBB data.  Bars present AUC values on the test set of UKBB (mean and 95\% CI from 10,000 bootstrap samples). Corresponding values are reported in [TODO]. For additional comparisons to C+T, SCT, PRS-CS, PRS-CS-auto and SBayesR \cite[]{prive2019making,ge2019polygenic,lloyd2019improved}, the reader is referred to figures 4 and S1 of \cite{prive2020ldpred2}.\label{fig:auc}}
\end{figure}

We then designed simulations where variants have different GWAS sample sizes, which is often the case when meta-analyzing GWAS from multiple cohorts without the same genome coverage.
Using [XXX] variants of chromosome 22, we simulate [BLABLA] where half of the variants use 100\% of the individuals for GWAS, one quarter uses only 80\%, and the remaining quarter uses only 60\% of the total GWAS sample size. We then run C+T, LDpred2-inf, LDpred2(-grid), LDpred2-auto, lassosum, lassosum2, PRS-CS and SBayesR by using either the true per-variant GWAS sample size, or the total sample size. Note that C+T does not use the sample size information, PRS-CS can only use one global sample size and SBayesR always diverged (so we discarded it).

[BLABLA ABOUT BAD SUMSTATS EXCEPT THE PGC's + ADD FIGURE ON QC FROM LDPRED2]

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\clearpage
%\vspace*{5em}

\section*{Code and results availability}

All code used for this paper is available at \url{https://github.com/privefl/paper-lassosum2/tree/master/code}.
We have extensively used R packages bigstatsr and bigsnpr \cite[]{prive2017efficient} for analyzing large genetic data, packages from the future framework \cite[]{bengtsson2020unifying} for easy scheduling and parallelization of analyses on the HPC cluster, and packages from the tidyverse suite \cite[]{wickham2019welcome} for shaping and visualizing results.
We have also used R package deming for fitting Deming regressions.

R packages bigstatsr and bigsnpr can be installed from GitHub and CRAN.
A tutorial on fitting penalized regressions with R package bigstatsr is available at \url{https://privefl.github.io/bigstatsr/articles/penalized-regressions.html}.
A tutorial on running LDpred2 with R package bigsnpr is available at \url{https://privefl.github.io/bigsnpr/articles/LDpred2.html}.


\section*{Acknowledgements}

Authors thank GenomeDK and Aarhus University for providing computational resources and support that contributed to these research results.
This research has been conducted using the UK Biobank Resource under Application Number 58024.

\section*{Funding}

F.P.\ and B.J.V.\ are supported by the Danish National Research Foundation (Niels Bohr Professorship to Prof. John McGrath).

\section*{Declaration of Interests}

%S.C.\ is a paid consultant to MyHeritage.
The other authors declare no competing interests.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\clearpage

\bibliographystyle{natbib}
\bibliography{refs}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\end{document}
